lib/include/jxl/codestream_header.h

```diff
@@ -314,6 +314,32 @@ typedef struct {
   /** Indicates this is the last animation frame.
    */
   JXL_BOOL is_last;
+
+  /** If coalescing is enabled, the following can be ignored; crop_x0/crop_y0 is
+     always zero, xsize/ysize is always equal to the image dimensions, and
+     blendmode is always 0 (kReplace). If coalescing is disabled, the layer has
+     dimensions xsize x ysize (which can be smaller or larger than the image
+     dimensions) and its top left corner is at position (crop_x0,crop_y0) -
+     these offsets can be negative.
+   */
+  /** Horizontal offset of the frame (can be negative)
+   */
+  int32_t crop_x0;
```

Comment:
This is ABI incompatible in the decoder so that would need to be noted and handled too (SOVERSION needs to be bumped).

crop_x0, crop_x1, xsize and ysize are useful even in the coalescing case with animation (GIF use-cases would be the obvious example). It would make sense to coordinate this change with that feature too, which is a feature required for djxl to transcode a gif (without blending all the frames to full-size frames).

Have you consider having these in a separate struct which you only get using these "advanced" modes?

Reply:
For GIF/APNG transcoding, you would use non-coalescing decoding, because otherwise you'd only get full-sized frames (and even if you would still get the crop offset and 'real' frame size, the pixel data would be after blending which is not what you want).

The crop coords/offset could still be useful in case of coalesced animation if you want to use the libjxl blending but still only want to refresh part of the screen. For that to work though, in principle what you need is not the cropbox of the last nonzero-duration frame, but a bounding box around all frames since the last nonzero-duration frame.

Separate struct could make sense, it might help for conceptual clarity but also if we want to do a lightweight libjxl_dec that doesn't include stuff like non-coalescing that isn't needed for viewing images/animations, it might make it easier to put these things behind a compile flag. Would you make this struct part of the JxlFrameHeader struct, or make it an entirely separate struct that can be accessed with a new function?

Reply:
For GIF/APNG animations you still need to get the coalesced frames, not the non-coalesced. Each frame can have layers, but that's not part of what you want to display.

Indeed cropped x0/y0 here would mean the crop in the coalesced API frames, not the internal zero-duration ones; that's the pixel data we are returning. The frame would not be "after blending", it would the frame that was meant to be as a second animation frame for example, even if it has alpha. That second frame could have been created with multiple layers, but that can still be all internal. Getting cropped frames when the output outside a given rectangle didn't change has applications in decoder applications (only refresh the portion of the animation that changed), not just in gimp/photoshop type of use-cases like the non-coalescing feature.

We want to support the coalesced use case for sure. Mixing "layers" and frames in the API is making all this more complex, which is why I don't think this is a good time to add the feature.

Reply:
Bumped the SOVERSION and moved the info to a separate struct. For the color channels it is always returned as part of the frame header, for the extra channels the blend info for those has to be fetched separately with a new function (to keep things fixed-size).